components/builder-core/src/api.rs

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::result;
+use std::str::FromStr;
+
+use hab_net::routing::Broker;
+use protocol::originsrv::{CheckOriginAccessRequest, CheckOriginAccessResponse, Origin,
+                          OriginChannel, OriginChannelCreate, OriginChannelGet, OriginGet,
+                          OriginPackage, OriginPackageGet, OriginPackageGroupPromote,
+                          OriginPackageIdent, OriginPackagePromote};
+use protocol::net::{ErrCode, NetError, NetOk};
+use protocol::scheduler::{Group, GroupGet, GroupState, Project, ProjectState};
+use protocol::sessionsrv::{Session, SessionCreate, SessionGet};
+
+use data_structures::UnsuccessfulJobGroupPromote;
+
+use error::{Error, Result};
+
+pub fn get_origin<T: ToString>(origin: T) -> result::Result<Option<Origin>, NetError> {
+    let mut conn = Broker::connect().unwrap();
+    let mut request = OriginGet::new();
+    request.set_name(origin.to_string());
+
+    match conn.route::<OriginGet, Origin>(&request) {
+        Ok(origin) => Ok(Some(origin)),
+        Err(err) => {
+            if err.get_code() == ErrCode::ENTITY_NOT_FOUND {
+                Ok(None)
+            } else {
+                Err(err)
+            }
+        }
+    }
+}
+
+pub fn check_origin_access<T: ToString>(
+    account_id: u64,
+    origin: T,
+) -> result::Result<bool, NetError> {
+    let mut conn = Broker::connect().unwrap();
+    let mut request = CheckOriginAccessRequest::new();
+    request.set_account_id(account_id);
+    request.set_origin_name(origin.to_string());
+    match conn.route::<CheckOriginAccessRequest, CheckOriginAccessResponse>(&request) {
+        Ok(response) => Ok(response.get_has_access()),
+        Err(err) => Err(err),
+    }
+}
+
+pub fn create_channel(origin: &str, channel: &str, session_id: u64) -> Result<(OriginChannel)> {
+    let origin_id = match get_origin(origin).map_err(Error::NetError)? {
+        Some(o) => o.get_id(),
+        None => {
+            debug!("Origin {} not found!", origin);
+            return Err(Error::OriginNotFound(origin.to_string()));
+        }
+    };
+
+    let mut conn = Broker::connect().unwrap();
```

Comment:
What's the cost of setting up a Broker connection? Wondering if piping a Broker through and reusing it might be nicer from that perspective, too.